Td blueprint fsm v1 - fix https://github.com/Vantiq/Blueprints/issues/12

realtdegen opened the pull request:
Remove spaces from client-name.

Fixes #12

realtdegen committed:
fixed import problems into VANTIQ 1.29 of email source and "Service Desk" pageset

Some resources could not be imported successfully. Encountered the following errors:

Failed to import sources email because of error: Request failed due to exception: java.lang.NumberFormatException: For input string: "[port]"
Failed to import document public/css/customcalendar.css because of error: {"data":null,"status":-1,"config":{"method":"POST","transformRequest":[null],"transformResponse":[null],"url":"https://dev.vantiq.com/api/v1/resources/documents","data":{},"headers":{"Authorization":"Bearer eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJaUlEybmJROHhvTHNwS3RBaE9mMllESENPeWVqa0tOVHhOSnlpT1RKeWZJIn0.eyJleHAiOjE2MDU5NjQ2MTIsImlhdCI6MTYwNTg3ODIxMiwiYXV0aF90aW1lIjoxNjA1ODc4MjEyLCJqdGkiOiJiNmQ1MjBjMy00MzM5LTRhMDktYTE5Yy0zZTg1ODUzZDdiODMiLCJpc3MiOiJodHRwczovL2Rldi52YW50aXEuY29tL2F1dGgvcmVhbG1zL2Rldi52YW50aXEuY29tIiwic3ViIjoiMDY1ZDRiYmYtZjQ4ZS00Y2E1LTk0NGItOTg5NjYxZWRmMTU0IiwidHlwIjoiQmVhcmVyIiwiYXpwIjoidmFudGlxU2VydmVyIiwic2Vzc2lvbl9zdGF0ZSI6ImM5MWJlODdiLWFjNGMtNDA1My1iMzVlLTlmYWE0NWYwZjBjYSIsImFjciI6IjEiLCJzY29wZSI6Im9wZW5pZCBwcm9maWxlIGVtYWlsIiwiZW1haWxfdmVyaWZpZWQiOnRydWUsIm5hbWUiOiJUaG9tYXMgRGVnZW4iLCJwcmVmZXJyZWRfdXNlcm5hbWUiOiJyZWFsdGRlZ2VuIiwiZ2l2ZW5fbmFtZSI6IlRob21hcyIsImZhbWlseV9uYW1lIjoiRGVnZW4iLCJlbWFpbCI6InRkZWdlbkB2YW50aXEuY29tIn0.dwkpZJlKu3LaiNCumLqxu2V8hC72o8Z5KhW3G3INBkTSyqCy9aKyzZci6XOEtgcSEsL9tC9ILXOn4owsQrHwxuflymsDEAB8Sg0Z_SNjUoLZom5oNatqqBz6OsUrzlo5LN8k2iE9z6XHXX878wowIztvmSi7pqaaiF_BYrtvTf-G3vcU3kF8Wf2n8DsIsANR7NiPDKbjapU8r5qtlyj7gVFfyycqY0DYDu-I9natGo6s1pOhujRBU9UqRRrrDp06VmRECrxxAUpZNXzwr-UYR_8PMzwoVhvzEQyGp1X3n-yMnP50uVa6Mc946GBk4tVUuGVYjZz1onkKhEx1bpD71A","X-Target-Namespace":"FieldServiceManagement","Accept":"application/json, text/plain, */*"}},"statusText":""}

realtdegen committed:
Update email.json

JakeSchmitz commented:
Does the project import for you in a clean namespace from this branch, and can you then run the ServiceDesk client successfully? I tried importing from this branch in a clean namespace and got the following error:

```
Exception: One or more custom assets failed to load: 
 ../docs/NS/bp_fsm/css/servicedesk.css
../docs/NS/bp_fsm/css/fullcalendar.css
../docs/NS/bp_fsm/css/customcalendar.css
../docs/NS/bp_fsm/css/slimselect.css
../docs/NS/bp_fsm/status/status_image_flow.css
../docs/NS/bp_fsm/js/moment.min.js
../docs/NS/bp_fsm/js/fullcalendar.js
../docs/NS/bp_fsm/js/slimselect.js
```

A lot of those documents appear to exist, but the paths are wrong.

rblumer4 commented:
the project FSM_CLIENT also needs to be updated to reference the new client name ServiceDesk (instead of Service Desk).

realtdegen commented:
> Does the project import for you in a clean namespace from this branch, and can you then run the ServiceDesk client successfully? I tried importing from this branch in a clean namespace and got the following error:
> 
> ```
> Exception: One or more custom assets failed to load: 
>  ../docs/NS/bp_fsm/css/servicedesk.css
> ../docs/NS/bp_fsm/css/fullcalendar.css
> ../docs/NS/bp_fsm/css/customcalendar.css
> ../docs/NS/bp_fsm/css/slimselect.css
> ../docs/NS/bp_fsm/status/status_image_flow.css
> ../docs/NS/bp_fsm/js/moment.min.js
> ../docs/NS/bp_fsm/js/fullcalendar.js
> ../docs/NS/bp_fsm/js/slimselect.js
> ```
> 
> A lot of those documents appear to exist, but the paths are wrong.

of course this is not working in the current branch, I‘ve fixed these bad assets referenced by multiple clients after the overall FSM solution got imported successfully first which is what this defect here is about....actually I‘ve thought to log these bad references to none existent assets only discovered at runtime not at import time as a new defect; the import agent should at least warn the user about existent stale asset references and not silently import these resources which will fail at runtime; I remember that Namir (or somebody else) is working on fixing all such bad asset references in 1.30 which is what I wanted to check this solution to import against next
BTW, I‘m still on 1.29 dev.vantiq.com where the import does not fail, I assume you are on 1.30 where these bad assets are caught at import ? if you are on 1.30 and get the warnings at import time about bad stale references to none existent assets, then I do not need to log this as a new defect and will wait for 1.30

realtdegen commented:
> the project FSM_CLIENT also needs to be updated to reference the new client name ServiceDesk (instead of Service Desk).

same like my previous comment, after the successful import the experienced user can add the ServiceDesk client from the namespace pool on its own to FSM_CLIENT project which is what I did trying to get this FSM solution after successful import up and running....I will update the whole package with this issue and the bad assets mentioned by Jake and commit to this outstanding pull request

realtdegen committed:
renamed Email Source "EmailSource" to original "email"

renamed EmailSource to email

realtdegen committed:
Revert "renamed Email Source "EmailSource" to original "email""

This reverts commit abeb85c11f4ee24c67fedf1ba8a0f8aade9cd3bf.

realtdegen committed:
Updated FSM_CLIENT and FSM_MOBILE_CLIENTS projects to be runnable after import

Resolved wrong referenced assets to mobile clilents in project FSM_MOBILE_CLIENTS and added new ServiceDesk client (without the space in between as in previous 'Service Desk' client name) to FSM_CLIENT ; as I've used the CLI's VCSSERVER to "sync project to VCS" aka my local GitHub branch, a lots of files got changed now

realtdegen commented:
please check, approve and merge; resolved with last commit (first commit and its revert was due to bad summary/comment linked to the commit) : wrong referenced assets to mobile clients in project FSM_MOBILE_CLIENTS and added new ServiceDesk client (without the space in between as in previous 'Service Desk' client name) to FSM_CLIENT ; 
info : as I've used the CLI's VCSSERVER to "sync project to VCS" aka my local GitHub branch, a lots of files got changed now compared to previous source files :
C:\VANTIQ\cli\vantiq-1.29.16>vantiq VCSSERVER
11/25 11:29:28 Server active (1.29.16)
11/25 11:41:57 Directory 'C:\VANTIQ\GitHub\Blueprints\Field Service Management' has been synced from project FSM_MOBILE_CLIENTS
11/25 11:42:15 Directory 'C:\VANTIQ\GitHub\Blueprints\Field Service Management' has been synced from project FSM_CLIENT

realtdegen committed:
resolved the mess by VCSSERVER in the FSM folder which created new folders for FSM_CLIENT and FSM_MOBILE_CLIENTS projects instead of following the namespace folder structure

the subject says it all....

realtdegen commented:
now it is all as supposed to be....the VCSSERVER export messed up the FSM folder structure.....

rblumer4 commented:
> now it is all as supposed to be....the VCSSERVER export messed up the FSM folder structure.....

Yes, the UI VCSSERVER export is by project, while the export in this Blueprint was done as a  CLI "whole namespace" export.  You have to pick one or the other and then stick with the same export-flavor for later exports, otherwise the directory and filenames won't match.
Looks like you've fixed that to get the matching file names.

rblumer4 commented:
> Does the project import for you in a clean namespace from this branch, and can you then run the ServiceDesk client successfully? I tried importing from this branch in a clean namespace and got the following error:
> 
> ```
> Exception: One or more custom assets failed to load: 
>  ../docs/NS/bp_fsm/css/servicedesk.css
> ../docs/NS/bp_fsm/css/fullcalendar.css
> ../docs/NS/bp_fsm/css/customcalendar.css
> ../docs/NS/bp_fsm/css/slimselect.css
> ../docs/NS/bp_fsm/status/status_image_flow.css
> ../docs/NS/bp_fsm/js/moment.min.js
> ../docs/NS/bp_fsm/js/fullcalendar.js
> ../docs/NS/bp_fsm/js/slimselect.js
> ```
> 
> A lot of those documents appear to exist, but the paths are wrong.

@realtdegen, it looks like these errors are occurring for @JakeSchmitz because there are TWO clients in the export now.  One under the directory "pagesets" and one under the directory "clients".   When we do a full import of the whole directory, the "pageset" one is overwriting the "clients" one.

"pagesets" is the old name for clients, and has been discontinued.  
I think you should delete that directory, and then this PR will be correct & merge-able.

(And the client will still need to be fixed-up after import, to replace "docs/NS" with "docs/\<namespaceName>".  But I assume those directions are somewhere in the Blueprint setup instructions.)

realtdegen commented:
the VCSSERVER should allow for both types of export directory structures, either project (as is) and new as ns export by CLI, I will log an Aha idea for this;
let me remove the stale clients below „pagesets“ folder, will do so with my next commit

rblumer4 commented:
Might be good to note the export structure in the README.md file.
e.g.  add this line to the end of the file
>Note: when updating this project in Github, use the CLI to export the files, not VCS.  The export format in the Github repository is in the CLI full-namespace export structure (and not the UI VCS per-project export structure).

realtdegen committed:
removal of duplicate client exports from legacy 'pagesets' folder

as subject

realtdegen commented:
as previous commit says, I've removed the duplicate clients in outversioned folder 'pagesets' and successfully imported the current set of FieldServiceManagement resources into a vanilla namespace on internal.vantiq.com (1.29); after the import I could start the ServiceDesk client in FSM_CLIENT project without runtime error pointing to non-existent custom assets and I could at least by MPI start some mobile clients from FSM_MOBILE_CLIENTS project who have complained before about bad paths/URLs to servicedeskmobile.css and other supplementary custom css stylesheets

realtdegen commented:
looks like I've missed your attention & availability for this before Thanksgiving holiday break....is it actually you - Roger - reviewing, approving and merging my branch into master ? or are you only reviewing and finally Jake is approving/merging ? I'm still not clear about ownership and responsibility especially who is engaged here in the Vantiq/Blueprints GitHub repository and responsible for

realtdegen committed:
Fix https://github.com/Vantiq/Blueprints/issues/19

updated Technicians type having bad property type as reported in https://github.com/Vantiq/Blueprints/issues/19

realtdegen committed:
New UI for FSM_MOBILE_CLIENTS enhancements from api.vantiq.com

the WorkOrders mobile client UI is much nicer and enhanced existent on api.vantiq.com which I've upgraded the current FSM code to

realtdegen commented:
the more I'm using the FSM solution and find things, the more I can put in this PR here.....

realtdegen committed:
Upgrade ServiceDesk client of FSM_CLIENT project

upgraded FSM_CLIENT especially the ServiceDesk web client to latest one as existent on api.vantiq.com

realtdegen committed:
added important note to update status_image_flow.css after import into new namespace accordingly

it took me a while to figure this out, you have to update css asset public/status/status_image_flow.css after import into a new namespace to match this new runtime environment for getting the full effect of this stylesheet

realtdegen commented:
it took me a while to figure out how to get these graphics at runtime on api.vantiq.com namespace "br_acmeWater" part of the work orders details page up and running in my local namespaces "FieldServiceManagement" on dev.vantiq.com and internal.vantiq.com....this is somehow hardcoded for your runtime namespace in CSSasset public/status/status_image_flow.css, I've had to exchange the reference to NS "br_acmeWater" with "FieldServiceManagement" at ten positions in custom css status_image_flow.css to get the effect from this CSS

Question to engineering : Do you know how to make this reference embedded in this CSS stylesheet relative wrt to the runtime and hence portable across different namespaces and VANTIQ environments without the need to update manually before/after its import into a new NS/runtime env ?

the user has to update this imported CSS asset at ten positions and have to re-point to the current execution namespace which I've added to the README.md

Note: To get the full effect at runtime in the ServiceDesk client Work Orders detail page, you need to accommodate
after import the referenced CSS custom asset public/status/status_image_flow.css to match your namespace at runtime
e.g. replace here for 'background : url' referenced namespace 'br_acmeWater' in between 'NS' and 'status' properties
with 'my_fsmblueprint_ns' at these 10 positions :

/* To use CSS - change the namespace name and folder where images are uploaded. Update 'left' positions. */
#list_image {
                background : url(../ui/docs/NS/br_acmeWater/status/list_image.png) no-repeat;
                position: absolute;
...
}
#list_gray {
                background : url(../ui/docs/NS/br_acmeWater/status/list_gray.png) no-repeat;
                position: absolute;
...
}
#person_image {
                background : url(../ui/docs/NS/br_acmeWater/status/person_image.png) no-repeat;
                position: absolute;
...
}
#person_gray {
                background : url(../ui/docs/NS/br_acmeWater/status/person_gray.png) no-repeat;
                position: absolute;
...
}
#enroute_image{
                background : url(../ui/docs/NS/br_acmeWater/status/enroute_image.png) no-repeat;
                position: absolute;
...
}
#enroute_gray{
                background : url(../ui/docs/NS/br_acmeWater/status/enroute_gray.png) no-repeat;
                position: absolute;
...
}
#technician_image{
                background : url(../ui/docs/NS/br_acmeWater/status/technician_image.png) no-repeat;
                position: absolute;
...
}
#technician_gray{
                background : url(../ui/docs/NS/br_acmeWater/status/technician_gray.png) no-repeat;
                position: absolute;
...
}
#checkbox_image{
                background : url(../ui/docs/NS/br_acmeWater/status/checkbox_image.png) no-repeat;
                position: absolute;
...
}
#checkbox_gray{
                background : url(../ui/docs/NS/br_acmeWater/status/checkbox_gray.png) no-repeat;
                position: absolute;
...
}

realtdegen commented:
find here the graphics from the ServiceDesk client's 'workorderpage' on the right coming with "Created/Assigned/En Route/On Site/Completed" timestamps requiring to update public/status/status_image_flow.css that I've referred to in the last commit with the note to README.md
 
![image](https://user-images.githubusercontent.com/50694969/100423184-e1188500-308b-11eb-8e47-14661400c572.png)

realtdegen committed:
Fix bad CSS asset references for servicemobileclient.css

Fix for GH issue https://github.com/Vantiq/Blueprints/issues/23 bad CSS asset references to servicemobileclient.css

realtdegen commented:
fixed bad CSS asset reference to servicedeskmobile.css in all mobile clients

one client has even been referring to this stylesheet located at https://dev.vantiq.com/ui/docs/NS/bp_sitsm_v2/css/servicedeskmobile.css

Edit Properties for Client 'Accept-Reject-Assignment'
Custom Assets
CSS Assets
https://dev.vantiq.com/ui/docs/NS/bp_sitsm_v2/css/servicedeskmobile.css

realtdegen committed:
added missing required procedures for ServiceDesk client

added missing required procedures cb.getAllCustomers / cb.getContactsByCustomerId / cb.getLocationByLocationId and cb.getLocationsByCustomerId for latest ServiceDesk client to resolved https://github.com/Vantiq/Blueprints/issues/22

realtdegen commented:
fixed missing cb.* procedures for ServiceDesk client with last commit https://github.com/Vantiq/Blueprints/issues/22

realtdegen committed:
more and more....after trying to fix the email notification issue found more problems with the collaboration

lot of additional thhings required to get the final TechnicianDispatch up and running like additional NLP and chatbot sources, updated email NotificationTemplates type, changes to procedure collab.scheduleLocationTracking and more.....

realtdegen committed:
the servicedeskmobile.css duplication story....to be continued

really need to figure out why this stylesheet is bouncing around so much...but do not want to update the original source on api.vantiq.com so duplicate this to have it around in any case.....

rblumer4 commented:
> looks like I've missed your attention & availability for this before Thanksgiving holiday break....is it actually you - Roger - reviewing, approving and merging my branch into master ? or are you only reviewing and finally Jake is approving/merging ? I'm still not clear about ownership and responsibility especially who is engaged here in the Vantiq/Blueprints GitHub repository and responsible for

I'm not sure who owns this FSM app, but I was leaving it to @JakeSchmitz to approve the PR.

Of course,  I'm not sure that Jake or I can really know if the changes make the app usable or not.
@pburma, @bruden01,  do you still use this Blueprint?
Can you review and vet the changes that have been made?

@realtdegen, you've changed the README to say to update the CSS in order to use it, but your instructions do not match the checked-in files.  For instance, you mention `br_acmeWater` in the README, but I do not see that string in any files in the branch.  Instead I see strings like:  `../ui/docs/NS/FieldServiceManagement/...`
Can you verify & update those instructions to match the strings that are checked-in?

rblumer4 commented on Field Service Management/README.md:
```diff
@@ -33,5 +33,66 @@ The FSM Blueprint is an application built to provide ideal solutions these chall
-If you need assistance with getting started please contact VANTIQ's technical support team. 
+If you need assistance with getting started, please contact VANTIQ's technical support team at support@vantiq.com .
 
+Note: To get the full effect at runtime in the ServiceDesk client Work Orders detail page, you need to accommodate
+after import the referenced CSS custom asset public/status/status_image_flow.css to match your namespace at runtime
```
might rephrase this
```suggestion
Note: To get the full effect at runtime in the ServiceDesk client Work Orders detail page, after import you need to update
the referenced CSS custom asset public/status/status_image_flow.css to match your namespace name.
```

rblumer4 commented on Field Service Management/README.md:
```diff
@@ -33,5 +33,66 @@ The FSM Blueprint is an application built to provide ideal solutions these chall
+
+/* To use CSS - change the namespace name and folder where images are uploaded. Update 'left' positions. */
+#list_image {
+                background : url(../ui/docs/NS/br_acmeWater/status/list_image.png) no-repeat;
```
in the checked-in code, these are all `FieldServiceManagement`    (and update `br_acmeWater` in lines above & below this)
```suggestion
                background : url(../ui/docs/NS/FieldServiceManagement/status/list_image.png) no-repeat;
```

rblumer4 reviewed (commented):
saw these inconsistencies, but haven't tried to import the app to verify it works.

realtdegen commented:
@rblumer4 I will update the README.md when I will be through the whole FSM solution e2e; the pointer to this .css stylesheet has been some commits before where I’m now; I’ve corrected the latest css asset making the url a bit neutral, however the question in between remains: how to put the doc references in the stylesheet relative to runtime NS making this asset 100% portable avoiding this tweak; 
the original issue where I’ve started with the FSM sources failing to import has been resolved in between but my current FSM branch is not fully functional (see latest two new issues that I‘ve logged); once resolved and FSM app is useable, PR could get considered for approval and going public if noone else will decline

rblumer4 commented:
> Question to engineering : Do you know how to make this reference embedded in this CSS stylesheet relative wrt to the runtime and hence portable across different namespaces and VANTIQ environments without the need to update manually before/after its import into a new NS/runtime env

@mjswan, @steve-langley, is this a known issue or Aha request?
@realtdegen, if not, can you add an Aha request for this feature?

JakeSchmitz commented:
The import now blows up with a red-screen error:

<img width="1367" alt="Screen Shot 2020-11-30 at 3 30 12 PM" src="https://user-images.githubusercontent.com/4064011/100661086-ce11e900-3307-11eb-942a-956b25fc339a.png">

This issue showed up locally against master as well on test and internal, so it's not a new problem. That error implies there is something wrong with the serialized form of the client in the new export. 

I'm going to need to look through the changelog since last week to figure out what happened.

realtdegen commented:
yes, getting the same exception in the IDE on 1.29 against internal.vantiq.com trying to import the latest td_blueprint_fsm_v1 branch
with the CLI I'm getting a NPE at import of WorkOrders client :
C:\VANTIQ\GitHub\Blueprints\Field Service Management>vantiq -s FSM_v1_int version
Client Version: 1.29.16
Server Version: 1.29.17
Server Build Date: Wed Nov 25 18:01:16 UTC 2020
Server Commit: e991639a2c23b8212d40f9fa1da07a7f5b7ab5e6
C:\VANTIQ\GitHub\Blueprints\Field Service Management>vantiq -s FSM_v1_int import
Could not read the contents of .\clients\WorkOrders\controllers\topbar\widgets\MenuButton17\onClick.js: Cannot set property 'onClick' on null object
ERROR detected Processing request: java.lang.NullPointerException: Cannot set property 'onClick' on null object
C:\VANTIQ\GitHub\Blueprints\Field Service Management>

realtdegen committed:
Fix

realtdegen committed:
Revert "Fix"

This reverts commit e82f88b4fdaaa2df3859c8313dc026baa190dab7.

realtdegen committed:
Fix for red screen at import of WorkOrders client

C:\VANTIQ\GitHub\Blueprints\Field Service Management>vantiq -s FSM_v1_int import
Could not read the contents of .\clients\WorkOrders\controllers\topbar\widgets\MenuButton17\onClick.js: Cannot set property 'onClick' on null object
ERROR detected Processing request: java.lang.NullPointerException: Cannot set property 'onClick' on null object

realtdegen commented:
the WorkOrders client source has been corrupted somehow (don't now how this could happen, but corruption has been existent); a fresh export of the WorkOrders clilent from latest working namespace and commit into td_blueprint_fsm_v1 branch has resolved the problem at import time as tested with an import of current sources into a brandnew NS FSM_v1 on internal.vantiq.com

rblumer4 commented:
> The import now blows up with a red-screen error:

@JakeSchmitz, are you looking at fixing the red-screen, so that the UI handles the problem, or at least prints a better error than a red-screen?  Otherwise I can file a bug on it.

JakeSchmitz commented:
I'm not planning on looking at the red screen. This is specifically a client import/export problem, and I don't really know how that works. My best guess is the code @realtdegen deleted in the recent commits was for a button that doesn't exist in the client, and importing the code without the corresponding button violated some invariant.

@steve-langley might know what's going on here, I think he worked on the client import/export handling. Without a reproduction of generating the corrupt export it might be tough to make any progress outside of bulletproofing the client import code to avoid the red screen (but still fail to import the client).

rblumer4 commented:
I added Vantiq/vantiq-issues#171 about the import stacktrace.

realtdegen committed:
Fix https://github.com/Vantiq/Blueprints/issues/26

Fix https://github.com/Vantiq/Blueprints/issues/26

main-min.js:194 Trying to add widget 'InputString962086' to hash and another object already has that name.
GraphicalObject.addToCanvasHash @ main-min.js:194

2main-min.js:194 Trying to add widget 'InputString962047' to hash and another object already has that name.
GraphicalObject.addToCanvasHash @ main-min.js:194

main-min.js:194 Trying to add widget 'InputString962043' to hash and another object already has that name.
GraphicalObject.addToCanvasHash @ main-min.js:194

main-min.js:194 Trying to add widget 'InputString962048' to hash and another object already has that name.
GraphicalObject.addToCanvasHash @ main-min.js:194

main-min.js:194 Trying to add widget 'Button99087' to hash and another object already has that name.
GraphicalObject.addToCanvasHash @ main-min.js:194

main-min.js:194 Trying to add widget 'StaticText99931' to hash and another object already has that name.
GraphicalObject.addToCanvasHash @ main-min.js:194

main-min.js:194 Trying to add widget 'VerticalLayout9831' to hash and another object already has that name.
GraphicalObject.addToCanvasHash @ main-min.js:194

realtdegen committed:
Fix for https://github.com/Vantiq/Blueprints/issues/28

TypeError: Error in 'select' success callback: Cannot read property 'supportStart' of undefined

realtdegen committed:
ommitted the runtime failure at // get the customer support status

update to last commit

realtdegen committed:
Fix Error in 'select' success callback: log.info is not a function

mixed up log.info with console.log

JakeSchmitz commented on Field Service Management/clients/Notify-Manager/controllers/client/customCode.js:
```diff
@@ -0,0 +1,63 @@
+//Upsert function
+function upsert(resource, data, client, callback){
```
This code shows up in multiple client's customCode section. It would be better to have this code in a single document that could be loaded as a custom asset in all clients that need to call upsert, update or select to keep the code in one place and avoid duplicating code.

realtdegen replied:
this is one of the functions from Brett's Client Builder framework which are getting new official helper functions within 1.30; I will rewrite the FSM Blueprint for 1.30 accordingly. remove the redundant duplicate code fragments and point to these new Client.xyz helper functions

![image](https://user-images.githubusercontent.com/50694969/102248721-c8202880-3f01-11eb-96a6-fd6dfa6ee45d.png)

JakeSchmitz commented on Field Service Management/clients/ScheduleArrivalDateTime/controllers/client/customCode.js:
```diff
@@ -0,0 +1,63 @@
+function update(resource, data, dbid, client, callback){
```
Same comment on custom code

realtdegen replied:
this is one of the functions from Brett's Client Builder framework which are getting new official helper functions within 1.30; I will rewrite the FSM Blueprint for 1.30 accordingly. remove the redundant duplicate code fragments and point to these new Client.xyz helper functions

![image](https://user-images.githubusercontent.com/50694969/102248721-c8202880-3f01-11eb-96a6-fd6dfa6ee45d.png)

JakeSchmitz commented on Field Service Management/clients/ServiceDesk/controllers/client/customCode.js:
```diff
@@ -0,0 +1,313 @@
+  );
+}
+
+//Upsert function
```
Same goes for duplicate code here

realtdegen replied:
this is one of the functions from Brett's Client Builder framework which are getting new official helper functions within 1.30; I will rewrite the FSM Blueprint for 1.30 accordingly. remove the redundant duplicate code fragments and point to these new Client.xyz helper functions

![image](https://user-images.githubusercontent.com/50694969/102248721-c8202880-3f01-11eb-96a6-fd6dfa6ee45d.png)

JakeSchmitz commented on Field Service Management/clients/ServiceDesk/controllers/page/assignTechnician/onClientStart.js:
```diff
@@ -0,0 +1,367 @@
+/* initialize the calendar
+-----------------------------------------------------------------*/
+this.data.calendar.init = function(){
+	$('#assignCalendar').fullCalendar({
```
Indentation in here looks screwed up, looks like it's because you mixed using tabs and spaces for indentation and github makes that look weird.

JakeSchmitz commented on Field Service Management/clients/ServiceDesk/controllers/page/contactpagepopup/onStart.js:
```diff
@@ -0,0 +1,30 @@
+var thisCopy = this;
+
+/* 
+var params = {where:{customerid:parameters.companyid}};
+select("Locations", params, client, function(response){
+	var json = response;
+	var locations = [{"value":"-select-", "label":"- select -"}];
+	$(json).each(function(index, item) {
+		$('#locationSelect').append($('<option>', { 
+			value: item.id,
+            text : item.name 
+         }));
+    });  
+});	
+*/
```
Delete commented out code or explain it in comments to the user.

JakeSchmitz commented on Field Service Management/clients/ServiceDesk/controllers/page/contactslistpopup/widgets/InputString962086/onChange.js:
```diff
@@ -0,0 +1,31 @@
+var params;
+var match = page.data.search;
+if (match){
+    params = {
+		where: { 
+            $and: [
+            	{companyid: page.data.companyid },        
+            	{ $or: [
+                	{firstname: {$regex: match, $options:"i"}},
+                	{lastname: {$regex: match, $options:"i"}},
+                	{department: {$regex: match, $options:"i"}}
+                
+            	]}
+        	]}        
+		};
```
indentation

JakeSchmitz commented on Field Service Management/clients/ServiceDesk/controllers/page/locationpage/onStart.js:
```diff
@@ -0,0 +1,48 @@
+// var thisCopy = this;
```
There is a lot of commented out code in this file...

JakeSchmitz commented on Field Service Management/clients/ServiceDesk/controllers/page/locationpage/widgets/DataTable80/onSelect.js:
```diff
@@ -0,0 +1,17 @@
+var myArray = page.data.locations.contacts;
+//var table = client.getWidget("contactsTable");
+//selectedrow = table.selectedRowObject;
```
remove

JakeSchmitz commented on Field Service Management/clients/ServiceDesk/controllers/page/locationpage/widgets/addlocationBtn/onClick.js:
```diff
@@ -0,0 +1,24 @@
+/*
```
delete the entire handler here?

JakeSchmitz reviewed (commented):
The ServiceDesk starts up for me successfully now. That's about as much of a review as I can do for this. 

There are a few general categories of feedback I have from my look through:
* Commonize duplicated code in a custom asset that can be added to all clients. This is a best practice we should emphasize.
* Remove dead code/commented out code, or at least explain in comments why it's there. 
* Remove duplicate clients, I noticed there is both an Accept_Reject_Assignment and a Accept-Reject-Assignment
* Cleanup indentation (mostly a result of mixing tabs and spaces I'm guessing)

In general, I'm starting to think this shouldn't be in the Blueprints repo, but should instead be in a repo managed by the Sales Engineering team.

realtdegen committed:
Removed legacy mobile client Accept-Reject-Assignment from FSM_MOBILE_CLIENTS

Removed legacy mobile client Accept-Reject-Assignment from FSM_MOBILE_CLIENTS

realtdegen committed:
Streamlined mobile notification client names and removed stale WorkOrders collaboration

clients\Notify-Manager -> clients\NotifyManager
clients\Accept_Reject_Assignment -> clients\AcceptRejectAssignment
collaborationtypes\WorkOrders -> Deleted

realtdegen committed:
Update FSM_MOBILE_CLIENTS.json

Update clients appearance in the project's IDE pane

realtdegen committed:
Added FSM_IOTWORKORDER project and related required resources

Added FSM_IOTWORKORDER project and related required resources

realtdegen committed:
Added FSM_DATAGENERATOR project and required resources

FSM_DATAGENERATOR is creating events for the FSM_IOTWORKORDER project

realtdegen committed:
housekeeping IoT.newTicket -> iot.newTicket

a little bit of clean.up

realtdegen committed:
Update collab.notifyCustomer.vail

Fix #24 https://github.com/Vantiq/Blueprints/issues/24

realtdegen commented:
> The ServiceDesk starts up for me successfully now. That's about as much of a review as I can do for this.
> 
> There are a few general categories of feedback I have from my look through:
> 
> * Commonize duplicated code in a custom asset that can be added to all clients. This is a best practice we should emphasize.
> * Remove dead code/commented out code, or at least explain in comments why it's there.
> * Remove duplicate clients, I noticed there is both an Accept_Reject_Assignment and a Accept-Reject-Assignment
> * Cleanup indentation (mostly a result of mixing tabs and spaces I'm guessing)
> 
> In general, I'm starting to think this shouldn't be in the Blueprints repo, but should instead be in a repo managed by the Sales Engineering team.

@JakeSchmitz , so far the available FSM Blueprint on GitHub does not import properly although public accessible, hence it is not a trustable project for new starters which is not good for VANTIQ's reputation. Can we discuss how to continue with this PR and how to share responsibilities in between field and engineering for such blueprints ? Our agreement could become frame for future releases of additional blueprints or similar demos.

realtdegen commented:
I'm going to import/test the latest FSM sources as in branch td_blueprint_fsm_v1 against VANTIQ 1.30 on test.vantiq.com

sfitts commented:
Just so everyone is on the same page regarding ownership/division of labor for this and the other blueprints:

The blueprints belong to the field. Engineering does not maintain them. Historically, engineering has helped the field with the mechanics of using github repositories but that is the extent of our involvement.

realtdegen commented:
@sfitts and @JakeSchmitz, I've understood blueprints being an entry-level product basis hand-crafted by the field for a by engineering supported solution after handed over to engineering; this is the case (or my understanding) for the Back-to-Work Accelerator blueprint, isn't it ?

sfitts commented:
> @sfitts and @JakeSchmitz, I've understood blueprints being an entry-level product basis hand-crafted by the field for a by engineering supported solution after handed over to engineering; this is the case (or my understanding) for the Back-to-Work Accelerator blueprint, isn't it ?

No.  Engineering does not currently take over the blueprints or perform long term maintenance of them.  The work on the Back-to-Work Accelerator was a little bit of an exception in that it involved a joint team and also led to some features in the product to make maintenance/development of the blueprints possible.  But going forward they are currently owned by the field.
